src/internal/Subscription.ts

```diff
@@ -53,180 +59,201 @@ export class Subscription implements SubscriptionLike {
   unsubscribe(): void {
     let errors: any[] | undefined;
 
-    if (this.closed) {
-      return;
-    }
-
-    let { _parentOrParents, _ctorUnsubscribe, _unsubscribe, _subscriptions } = (this as any);
+    if (!this.closed) {
+      this.closed = true;
 
-    this.closed = true;
-    this._parentOrParents = null;
-    // null out _subscriptions first so any child subscriptions that attempt
-    // to remove themselves from this subscription will noop
-    this._subscriptions = null;
+      const { _parents, _singleParent, _teardowns } = this;
+      this._parents = this._singleParent = this._teardowns = null;
+      let { _ctorUnsubscribe, _unsubscribe } = this as any;
 
-    if (_parentOrParents instanceof Subscription) {
-      _parentOrParents.remove(this);
-    } else if (_parentOrParents !== null) {
-      for (let index = 0; index < _parentOrParents.length; ++index) {
-        const parent = _parentOrParents[index];
-        parent.remove(this);
+      // Remove this from it's parents.
+      if (_singleParent) {
+        this._singleParent = null;
+        _singleParent.remove(this);
       }
-    }
 
-    if (isFunction(_unsubscribe)) {
-      // It's only possible to null _unsubscribe - to release the reference to
-      // any teardown function passed in the constructor - if the property was
-      // actually assigned in the constructor, as there are some classes that
-      // are derived from Subscriber (which derives from Subscription) that
-      // implement an _unsubscribe method as a mechanism for obtaining
-      // unsubscription notifications and some of those subscribers are
-      // recycled. Also, in some of those subscribers, _unsubscribe switches
-      // from a prototype method to an instance property - see notifyNext in
-      // RetryWhenSubscriber.
-      if (_ctorUnsubscribe) {
-        (this as any)._unsubscribe = undefined;
-      }
-      try {
-        _unsubscribe.call(this);
-      } catch (e) {
-        errors = e instanceof UnsubscriptionError ? flattenUnsubscriptionErrors(e.errors) : [e];
+      if (_parents) {
```

Comment:
nitpick: This could be an `else if` as the implementation is guaranteeing that `_parents` will be `null` if `_singleParent` is non-`null`. IMO, this is significant not just for perf, but also to aid a reader's understanding: if the guarantee is made, IMO, the code paths in the implementation should reflect that - rather than suggest the guarantee isn't here.